engine/lib/phases/phase_rewrite_control_flow.ml

```diff
@@ -11,31 +14,97 @@ module Make (F : Features.T) =
 
       open Ast.Make (F)
       module U = Ast_utils.Make (F)
+      module M = Ast_builder.Make (F)
       module Visitors = Ast_visitors.Make (F)
 
       module Error = Phase_utils.MakeError (struct
         let ctx = Diagnostics.Context.Phase phase_id
       end)
 
-      let has_return =
+      let has_cf =
         object (_self)
           inherit [_] Visitors.reduce as super
           method zero = false
           method plus = ( || )
 
+          method! visit_expr' break_continue e =
+            match e with
+            | Return _ -> true
+            | (Break _ | Continue _) when break_continue -> true
+            | _ -> super#visit_expr' break_continue e
+        end
+
+      let loop_return_type =
+        object (_self)
+          inherit [_] Visitors.reduce as super
+          method zero = (U.unit_typ, None)
+          method plus l r = if [%eq: ty] (fst l) U.unit_typ then r else l
+
           method! visit_expr' () e =
-            match e with Return _ -> true | _ -> super#visit_expr' () e
+            match e with
+            | Return { e; witness; _ } -> (e.typ, Some witness)
+            | _ -> super#visit_expr' () e
         end
 
       let rewrite_control_flow =
         object (self)
           inherit [_] Visitors.map as super
 
-          method! visit_expr () e =
+          method! visit_expr in_loop e =
+            let loop_with_return (loop : expr) stmts_after final pat =
+              let return_type, witness = loop_return_type#visit_expr () loop in
+
+              let typ =
+                U.M.ty_control_flow ~continue_type:loop.typ
+                  ~break_type:return_type
+              in
+              let loop = { loop with typ } in
+              let span = loop.span in
+              let id = U.fresh_local_ident_in [] "ret" in
+              let module MS = (val U.M.make span) in
+              let mk_cf_pat = U.M.pat_Constructor_CF ~span ~typ in
+              let return_expr =
+                let inner_e = MS.expr_LocalVar ~typ:return_type id in
+                match witness with
+                | Some witness ->
+                    MS.expr_Return ~typ:return_type ~witness ~inner_e
+                | None -> inner_e
+              in
+              let arms =
+                [
+                  MS.arm
+                    (mk_cf_pat `Break (U.make_var_pat id typ span))
+                    return_expr;
+                  MS.arm (mk_cf_pat `Continue pat)
+                    (U.make_lets stmts_after final |> self#visit_expr in_loop);
+                ]
+              in
+              MS.expr_Match ~scrutinee:loop ~arms ~typ:return_type
+            in
             match e.e with
-            | _ when not (has_return#visit_expr () e) -> e
-            (* Returns in loops will be handled by issue #196 *)
-            | Loop _ -> e
+            (* This is supposed to improve performance but it might actually make it worse in some cases *)
```

Comment:
Can you file an issue and reference it there so that we track that?

Reply:
https://github.com/hacspec/hax/issues/1033